### DISCUSSION TITLE
Configurable ZwaveJS log location

### DISCUSSION THREAD

**Discussion by justindthomas** [zwave-js/zwave-js-ui#2166] [Ideas]
Would folks be interested in a PR if I added the ability to specify the ZwaveJS log location? I'm running `zwavejs2mqtt` on a dedicated Raspberry Pi with an SD card and I'd like to leave logging pretty noisy, but want to do it to a ramdisk to reduce writes on the SD card.

I've never used Vue (I'm familiar with React) but it looks straightforward enough and I think I can get a PR in if folks would be open to the change.

**Comment by robertsLando** [comment]
@justindthomas I would suggest to change store directory in order to store everything on that location and not onlly logs as most of writes also happens on cache files used by zwavejs and them are stored there too

**Reply by justindthomas** [reply]
Okay, thanks. I'll think about how best to do that Probably not in a ramdiak like I was thinking since I'd lose all state on reboot. I suppose I could take periodic snapshots.

I don't know if it's a huge concern anyway with just this one service running on the Pi. I use USB SSD drives on other Pis to avoid the worry, but I wanted to keep this one simple with just the one dongle for the Z-Stick and no powered USB hub, etc.

**Reply by kpine** [reply]
This is true, but reducing the number of writes to sensitive storage would still be a benefit. There's also a big difference between cache files and logs. Cache files need to be persisted in order for restarts to work properly. Logs can be discarded without an impact. Putting the store directory on a ramdisk is not a good idea.

**Comment by kpine** [comment]
An [environment variable](https://github.com/kpine/zwave-js-server-docker/blob/96caf7a3370d0cf563aee07583fe21f6182f87ed/options.js#L6-L8) to override the default setting would be an easy, low-impact fix for advanced users.

A more invasive workaround that doesn't require any software changes is to define the Z-Wave JS logging options yourself in `settings.json`. This means taking control of the settings file and not letting zwavejs2mqtt modify it. In my deployment of z2m, I hand created the `settings.json` file and map it into the container as read-only. You can then customize the node-zwave-js logging settings, or any other z-wave settings.

Instead of:
```json
  "zwave": {
    "port": "/dev/zwave",
    "commandsTimeout": 30,
    "logLevel": "debug",
    "logEnabled": true,
    "securityKeys": {
      "S2_Unauthenticated": ""
    },
    "deviceConfigPriorityDir": "/usr/src/app/store/config",
    "logToFile": true,
    "serverEnabled": true,
    "enableSoftReset": true,
    "enableStatistics": false,
    "serverPort": 3000,
    "disclaimerVersion": 1
  }
```
you can use the following:
```json
  "zwave": {
    "port": "/dev/zwave",
    "commandsTimeout": 30,
    "securityKeys": {
      "S2_Unauthenticated": ""
    },
    "deviceConfigPriorityDir": "/usr/src/app/store/config",
    "serverEnabled": true,
    "enableSoftReset": true,
    "enableStatistics": false,
    "serverPort": 3000,
    "disclaimerVersion": 1,
    "options": {
      "logConfig": {
        "enabled": true,
        "level": "debug",
        "logToFile": true,
        "filename": "/logs/zwavejs_%DATE%.log",
        "forceConsole": true
      }
    }
  }
```

zwavejs2mqtt allows you to define Z-Wave JS options directly in the `zwave.options` key. z2m will merge your options with its defaults. Unfortunately, it does not deep copy the objects, so you have to manually define the entire [`logConfig`](https://zwave-js.github.io/node-zwave-js/#/api/driver?id=logconfig) object, instead of just overriding the filename.

With the above settings, logs will be written to the container's `/logs` directory. You can map a tmpfs or anything to `/logs` from docker, or change the setting to a directory appropriate for your deployment. In this example, I'd have the following volumes in my docker compose file:

```yaml
    volumes:
      - ./store:/usr/src/app/store
      - ./logs:/logs
      - ./settings.json:/usr/src/app/store/settings.json
```

Of course, by controlling the settings file you have the burden of keeping up with any new options that are available, or handling any changes to the config schema.

**Reply by justindthomas** [reply]
Thanks for the information! An environment variable makes sense to me; I'll write something up and submit a PR.

**Comment by justindthomas** [comment]
PR submitted here: https://github.com/zwave-js/zwavejs2mqtt/pull/2174
